mdata/aggmetric.go

```diff
@@ -163,25 +163,38 @@ func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSp
 				badConsolidator.Inc()
 				return Result{}, err
 			case consolidation.Cnt:
-				agg = a.cntMetric
+				agg = aggregator.cntMetric
 			case consolidation.Lst:
-				agg = a.lstMetric
+				agg = aggregator.lstMetric
 			case consolidation.Min:
-				agg = a.minMetric
+				agg = aggregator.minMetric
 			case consolidation.Max:
-				agg = a.maxMetric
+				agg = aggregator.maxMetric
 			case consolidation.Sum:
-				agg = a.sumMetric
+				agg = aggregator.sumMetric
 			default:
 				err := fmt.Errorf("internal error: AggMetric.GetAggregated(): unknown consolidator %q", consolidator)
 				log.Errorf("AM: %s", err.Error())
 				badConsolidator.Inc()
 				return Result{}, err
 			}
 			if agg == nil {
-				return Result{}, fmt.Errorf("Consolidator %q not configured", consolidator)
+				return Result{}, fmt.Errorf("consolidator %q not configured", consolidator)
+			}
+			result, err := agg.Get(from, to)
+			if err != nil {
+				return Result{}, err
 			}
-			return agg.Get(from, to)
+			futurePoints := []schema.Point{}
+			if a.rob != nil {
```

Comment:
I think `a.RLock()` may be enough like line 226 in `AggMetric.Get()` as it's just reading?
Not quite sure about the `aggregator` as there's a comment above `// no lock needed cause aggregators don't change at runtime`. If it's causing problems it may be another issue? To my understanding `aggregator.xxxMetric` has its own lock so it should be fine.

Reply:
If it's just for the rob, then indeed RLock will suffice, not Lock.

That comment refers to accessing a.aggregators itself - the slice with pointers.  But I'm talking about the actual aggregators being pointed to. I think it has a few fields that we unsafely read and write. It's hard to believe we would have had such an obvious bug for years, so let me look into it .

Reply:
i think what we need is:
* https://github.com/grafana/metrictank/pull/2025, this should also clarify things
* a RLock here around the rob acess and the Foresee() call.